Update WaitForStableCluster to wait for only system pods to exist

damemi opened the pull request:
**What type of PR is this?**
/kind flake

**What this PR does / why we need it**:
When waiting for a stable cluster, some pods still needed to be deleted which affects e2e test assumptions about the number of pods on a cluster. This assumes that a stable cluster should have only system pods

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #84787 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

damemi commented:
/cc @ahg-g

ahg-g commented on test/e2e/framework/util.go:
```diff
@@ -2405,10 +2405,17 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
-	for len(currentlyNotScheduledPods) != 0 {
+
+	// Wait for all pending pods to be scheduled, and for only system pods to be scheduled
+	for len(currentlyNotScheduledPods) != 0 && len(scheduledPods) - len (systemPods.Items) != 0 {
```
use a variable, something like ```numNonSystemPods := len(scheduledPods) - len (systemPods.Items)```

Huang-Wei replied:
Is it possible that some system Pods are not scheduled? For example:

- currentlyNotScheduledPods: 2 regular Pods, 1 system Pods
- scheduledPods: 1 regular Pod, 2 system Pod
- systemPods: 3 Pods (1 not scheduled and 2 scheduled)

In this case, len(scheduledPods) - len (systemPods.Items) == 0.

damemi replied:
@Huang-Wei true, this probably isn't a great check. Maybe something more explicit like a separate loop that just waits for len(allNamespacesPods) == len(systemPods)?

ahg-g replied:
you can do this:

```
scheduledSystemPods, currentlyNotScheduledSystemPods = e2epod.GetPodsScheduled(masterNodes, systemPods)
```

and  the check would be```currentlyNotScheduledSystemPods != 0 || len(systemPods.Items) != len(allPods.Items)```

damemi replied:
@ahg-g updated with this check

ahg-g commented on test/e2e/framework/util.go:
```diff
@@ -2405,10 +2405,17 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 	scheduledPods, currentlyNotScheduledPods := e2epod.GetPodsScheduled(masterNodes, allPods)
-	for len(currentlyNotScheduledPods) != 0 {
+
+	// Wait for all pending pods to be scheduled, and for only system pods to be scheduled
```
update the comment: ```wait for system pods to be scheduled, and for pods in all other namespaces to be deleted```

neolit123 commented on test/e2e/framework/util.go:
```diff
@@ -2405,10 +2405,17 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 	}
 	allPods.Items = currentPods
+	systemPods, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(metav1.ListOptions{})
+	ExpectNoError(err)
```
please annotate this error with a description.
[1]

damemi replied:
done

neolit123 commented on test/e2e/framework/util.go:
```diff
@@ -2405,10 +2405,17 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 		time.Sleep(2 * time.Second)
 
+		systemPods, err = c.CoreV1().Pods(metav1.NamespaceSystem).List(metav1.ListOptions{})
+		ExpectNoError(err)
```
[1]

damemi replied:
done

neolit123 commented on test/e2e/framework/util.go:
```diff
@@ -2418,6 +2425,7 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 			break
 		}
 	}
+
```
please remove this empty line.

neolit123 commented on test/e2e/framework/util.go:
```diff
@@ -2405,10 +2405,17 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 
 	}
 	allPods.Items = currentPods
+	systemPods, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(metav1.ListOptions{})
```
> allPods, err := c.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})

this above is already problematic in case of thousands of pods.

damemi replied:
@neolit123 could you clarify how I could fix this? It seems this function is already dependent on polling all of the pods in the cluster. I could pull the system pods from that query above, but the system pod query is likely to usually be pretty small, right?

ahg-g replied:
@neolit123  doesn't the list of pods already exist in the informer cache? also the scheduler tests which uses this function does not really create thousands of pods.

neolit123 replied:
the system pods should be fine, but "all the pods" can block for a while in super-clusters and i wouldn't be surprised that nobody in the wild would run our test suite on such clusters because of that (possibly other reasons too). it feels to me this logic should reworked.

neolit123 replied:
@ahg-g the cache can help.
ok so this function is only used under test/e2e/scheduling/predicates.go
this means that this function has to be moved under `test/e2e/scheduling/` ideally.

i was under the impression that given this is part of the framework we are using it elsewhere which doesn't make sense.

neolit123 replied:
cc @oomichi for an opinion about the move.

ahg-g replied:
yes, this seems to be used only in the scheduler e2e tests, I agree it should be moved under ```test/e2e/scheduling```

neolit123 commented:
/assign @timothysc @johnSchnake 
for more eyes on this change.

ahg-g commented on test/e2e/framework/util.go:
```diff
@@ -2406,11 +2406,19 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
 		allPods, err := c.CoreV1().Pods(metav1.NamespaceAll).List(metav1.ListOptions{})
-		ExpectNoError(err)
+		ExpectNoError(err, "listing all pods in all namespaces while waiting for stable cluster")
 		scheduledPods, currentlyNotScheduledPods = e2epod.GetPodsScheduled(masterNodes, allPods)
```
this needs to change to 
```
scheduledSystemPods, currentlyNotScheduledSystemPods := e2epod.GetPodsScheduled(masterNodes, systemPods)
```

damemi replied:
Should we include the logic from above to filter out succeeded and failed pods inside this loop then too? (https://github.com/kubernetes/kubernetes/blob/ed7a68f3bdb1cdbdc51d0ba11b8c53a8dd29cd62/test/e2e/framework/util.go#L2399-L2404) It seems odd that it isn't included already, but if we're making `len(allPods.Items)` part of our condition checking we should probably sort out if that is needed

ahg-g replied:
moving everything to inside the loop sounds reasonable to me.

ahg-g commented on test/e2e/framework/util.go:
```diff
@@ -2406,11 +2406,19 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
+	_, currentlyNotScheduledSystemPods := e2epod.GetPodsScheduled(masterNodes, systemPods)
+
+	// Wait for system pods to be scheduled, and for pods in all other namespaces to be deleted
+	for len(currentlyNotScheduledPods) != 0 && (len(currentlyNotScheduledSystemPods) != 0 || len(systemPods.Items) != len(allPods.Items)) {
```
why do we need ```len(currentlyNotScheduledPods) != 0 &&```? I think the remaining part should be sufficient: it tests that all system pods are scheduled, and that all scheduled pods are system pods.

damemi replied:
True, updated

ahg-g commented on test/e2e/framework/util.go:
```diff
@@ -2406,11 +2406,19 @@ func WaitForStableCluster(c clientset.Interface, masterNodes sets.String) int {
-	for len(currentlyNotScheduledPods) != 0 {
+	systemPods, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(metav1.ListOptions{})
+	ExpectNoError(err, "listing all pods in kube-system namespace while waiting for stable cluster")
+	_, currentlyNotScheduledSystemPods := e2epod.GetPodsScheduled(masterNodes, systemPods)
```
scheduledSystemPods (and return that at the end)

damemi replied:
updated

neolit123 commented on test/e2e/framework/util.go:
```diff
@@ -2317,36 +2317,16 @@ func retryCmd(command string, args ...string) (string, string, error) {
-		}
-
+// ListNamespaceEvents lists the events in the given namespace.
+func ListNamespaceEvents(c clientset.Interface, ns string) error {
```
there is already DumpEventsInNamespace.
you can either:
- use DumpEventsInNamespace with it's sorting option
- add optional sorting to DumpEventsInNamespac as a boolean flag or a sort.Interface:
https://golang.org/pkg/sort/#Interface

damemi replied:
This diff is an error in rebasing; I did not add this function... fixing...

neolit123 commented:
/approve
for framework.

please squash your commits.

ahg-g commented on test/e2e/scheduling/framework.go:
```diff
@@ -16,9 +16,68 @@ limitations under the License.
+		}
+
+	}
+	allPods.Items = currentPods
```
can you factor this out to a function, perhaps ```allPods := getAllPods()``` and use it in both places.

ahg-g commented on test/e2e/scheduling/framework.go:
```diff
@@ -16,9 +16,68 @@ limitations under the License.
+	allPods.Items = currentPods
+	systemPods, err := c.CoreV1().Pods(metav1.NamespaceSystem).List(metav1.ListOptions{})
+	framework.ExpectNoError(err, "listing all pods in kube-system namespace while waiting for stable cluster")
+	scheduledSystemPods, currentlyNotScheduledSystemPods := e2epod.GetPodsScheduled(masterNodes, systemPods)
```
ditto, factor out to a function that returns two ```int``` variables and use it in both places, inside and outside the loop: 

```
scheduledSystemPods, notScheduledSystemPods := getSystemPods()
allSystemPods := scheduledSystemPods + notScheduledSystemPods
```

and replace ```len(systemPods.Items) ``` with ```allSystemPods```

damemi committed:
Update e2e framework WaitForStableCluster function

Move WaitForStable cluster to test/e2e/scheduling and update it to wait for only system pods to be ready in a stable cluster.

damemi commented:
@ahg-g updated with those functions factored out, please take a look again. I also squashed

ahg-g commented on test/e2e/scheduling/framework.go:
```diff
@@ -16,9 +16,69 @@ limitations under the License.
+	framework.ExpectNoError(err, "listing all pods in kube-system namespace while waiting for stable cluster")
+	scheduledSystemPods, currentlyNotScheduledSystemPods := e2epod.GetPodsScheduled(masterNodes, systemPods)
+	return len(scheduledSystemPods), len(currentlyNotScheduledSystemPods)
+}
```
this is a lint issue

ahg-g reviewed (commented):
lgtm, but you need to fix the lint issue

ahg-g commented:
/lgtm
/approve

ahg-g commented:
/retest
